lib/datalog/access.go

```diff
@@ -0,0 +1,366 @@
+// +build roletester
+
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package datalog
+
+// #cgo LDFLAGS: -Lroletester/target/release -lrole_tester
+// #include <stdio.h>
+// #include <stdlib.h>
+// extern int process_access(unsigned char *f, unsigned char *res, size_t input_len, size_t output_len);
+// extern void deallocate_rust_buffer(unsigned char *res, size_t len);
+import "C"
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/trace"
+)
+
+// NodeAccessRequest defines a request for access for a specific user, login, and node.
+type NodeAccessRequest struct {
+	Username  string
+	Login     string
+	Node      string
+	Namespace string
+}
+
+// NodeAccessResponse defines all interpreted facts from the input facts.
+type NodeAccessResponse struct {
+	Accesses        Facts
+	facts           Facts
+	mappings        map[string]uint32
+	reverseMappings map[uint32]string
+	errorStrings    map[int]string
```

Comment:
Strictly speaking, godocs on struct fields are optional, but I've found that they're very helpful for readers. So I would add godocs on these (even internal fields) to explain what they are / for.